Enabled buffs from Mastery

nerzhei opened the pull request:
- Added a configuration option to indicate the percentage increase in weapon stats per 100% mastery.
- Weapon mastery increases every offensive stat of the weapon.

aontas committed:
Update release.yml

aontas committed:
Update release.yml

aontas committed:
Update release.yml

nerzhei committed:
added way to configure leveling messages

nerzhei committed:
changed templates section

nerzhei committed:
removed unused using directive

nerzhei committed:
Merge remote-tracking branch 'origin/main' into develop

aontas committed:
Update thunderstore.toml

aontas committed:
Merge branch 'main' into develop

nerzhei committed:
Merge branch 'aontas:main' into main

nerzhei committed:
resolved merge conflicts

nerzhei committed:
addressed feedback

nerzhei committed:
removed unused usings

nerzhei committed:
Merge remote-tracking branch 'origin/main' into develop

nerzhei committed:
Merge remote-tracking branch 'origin/main' into develop

nerzhei committed:
enabled weapon mastery buffs

nerzhei committed:
enabled weapon mastery buffs

aontas commented on Extensions/UnitStatTypeExtensions.cs:
```diff
@@ -0,0 +1,42 @@
+﻿using ProjectM;
```
This should use something more like this so that we have more future support for other stat improvements:

```
using ProjectM;

namespace XPRising.Extensions;

public static class UnitStatTypeExtensions
{
    private enum Category
    {
        Offensive,
        Defensive,
        Resource,
        Other
    }

    private static Category StatCategory(this UnitStatType unitStatType)
    {
        switch (unitStatType)
        {
            case UnitStatType.PhysicalPower:
            case UnitStatType.SiegePower:
            case UnitStatType.CooldownRecoveryRate:
            case UnitStatType.SpellPower:
            case UnitStatType.PhysicalLifeLeech:
            case UnitStatType.SpellLifeLeech:
            case UnitStatType.PhysicalCriticalStrikeChance:
            case UnitStatType.PhysicalCriticalStrikeDamage:
            case UnitStatType.SpellCriticalStrikeChance:
            case UnitStatType.SpellCriticalStrikeDamage:
            case UnitStatType.AttackSpeed:
            case UnitStatType.DamageVsUndeads:
            case UnitStatType.DamageVsHumans:
            case UnitStatType.DamageVsDemons:
            case UnitStatType.DamageVsMechanical:
            case UnitStatType.DamageVsBeasts:
            case UnitStatType.DamageVsCastleObjects:
            case UnitStatType.DamageVsVampires:
            case UnitStatType.DamageVsLightArmor:
            case UnitStatType.DamageVsVBloods:
            case UnitStatType.DamageVsMagic:
            case UnitStatType.PrimaryAttackSpeed:
            case UnitStatType.PrimaryLifeLeech:
            case UnitStatType.PrimaryCooldownModifier:
            case UnitStatType.BonusPhysicalPower:
            case UnitStatType.BonusSpellPower:
            case UnitStatType.SpellCooldownRecoveryRate:
            case UnitStatType.WeaponCooldownRecoveryRate:
            case UnitStatType.UltimateCooldownRecoveryRate:
            case UnitStatType.MinionDamage:
                return Category.Offensive;
            case UnitStatType.MaxHealth:
            case UnitStatType.PhysicalResistance:
            case UnitStatType.FireResistance:
            case UnitStatType.HolyResistance:
            case UnitStatType.SilverResistance:
            case UnitStatType.SunChargeTime:
            case UnitStatType.SunResistance:
            case UnitStatType.GarlicResistance:
            case UnitStatType.SpellResistance:
            case UnitStatType.Radial_SpellResistance:
            case UnitStatType.PassiveHealthRegen:
            case UnitStatType.ResistVsUndeads:
            case UnitStatType.ResistVsHumans:
            case UnitStatType.ResistVsDemons:
            case UnitStatType.ResistVsMechanical:
            case UnitStatType.ResistVsBeasts:
            case UnitStatType.ResistVsCastleObjects:
            case UnitStatType.ResistVsVampires:
            case UnitStatType.ImmuneToHazards:
            case UnitStatType.HealthRecovery:
            case UnitStatType.PvPResilience:
            case UnitStatType.CCReduction:
            case UnitStatType.DamageReduction:
            case UnitStatType.HealingReceived:
            case UnitStatType.SilverCoinResistance:
            case UnitStatType.ShieldAbsorb:
                return Category.Defensive;
            case UnitStatType.MovementSpeed:
            case UnitStatType.EnergyGain:
            case UnitStatType.MaxEnergy:
            case UnitStatType.Vision:
            case UnitStatType.ReducedResourceDurabilityLoss:
            case UnitStatType.FallGravity:
            case UnitStatType.BloodDrain:
            case UnitStatType.BloodEfficiency:
            case UnitStatType.InventorySlots:
                return Category.Other;
            case UnitStatType.ResourcePower:
            case UnitStatType.ResourceYield:
            case UnitStatType.DamageVsWood:
            case UnitStatType.DamageVsMineral:
            case UnitStatType.DamageVsVegetation:
                return Category.Resource;
            default:
                return Category.Other;
        }
    } 
    public static bool IsOffensiveStat(this UnitStatType unitStatType)
    {
        return unitStatType.StatCategory() == Category.Offensive;
    }
    
    public static bool IsDefensiveStat(this UnitStatType unitStatType)
    {
        return unitStatType.StatCategory() == Category.Defensive;
    }
    
    public static bool IsResourceStat(this UnitStatType unitStatType)
    {
        return unitStatType.StatCategory() == Category.Resource;
    }
}
```

nerzhei replied:
Updated.

aontas commented on Hooks/BuffHook.cs:
```diff
@@ -76,6 +76,9 @@ private static void SpawnPrefix(ModifyUnitStatBuffSystem_Spawn __instance)
         foreach (var entity in entities)
         {
+            //not sure if this goes here
+            if (Plugin.WeaponMasterySystemActive) WeaponMasterySystem.ApplyWeaponBuffs(entity, entityManager);
```
Don't do this here. It should be done as a part of the `ApplyBuffs` so that we do not clash with other mods.

nerzhei replied:
Done.

aontas commented on Systems/WeaponMasterySystem.cs:
```diff
@@ -349,5 +353,67 @@ public static MasteryType WeaponToMasteryType(WeaponType weapon)
+                var statBuff = buffer[i];
+                if (statBuff.StatType.IsOffensiveStat())
+                {
+                    statBuff.Value = (float)(statBuff.Value * (1 + statsIncrease));
```
We don't really want to do it this way, as there is a potential for reapplying the same multiplier multiple times.

aontas commented on Extensions/PrefabExtensions.cs:
```diff
@@ -0,0 +1,79 @@
+﻿using ProjectM;
```
We shouldn't need this file. See the comments in the overview.

nerzhei replied:
Removed.

aontas reviewed (changes requested):
Need to add this to the EquipmentSystemHook.cs to ensure the appropriate buff is reapplied when equipping/unequipping weapons:

```

    
    /*
     * These WeaponLevelSystem Postfix patches attempt to ensure that the weapon mastery buffs get reapplied correctly.
     */
    
    [HarmonyPostfix]
    [HarmonyPatch(typeof(WeaponLevelSystem_Spawn), nameof(WeaponLevelSystem_Spawn.OnUpdate))]
    private static void WeaponSpawnPostfix(WeaponLevelSystem_Spawn __instance)
    {
        if (Plugin.ShouldApplyBuffs)
        {
            Plugin.Log(LogSystem.Buff, LogLevel.Info, "WeaponLevelSystem spawn POST");
            var entityManager = __instance.EntityManager;
            var entities = __instance.__query_1111682356_0.ToEntityArray(Allocator.Temp);
            foreach (var entity in entities)
            {
                if (!entityManager.TryGetComponentData<EntityOwner>(entity, out var entityOwner) ||
                    !entityManager.TryGetComponentData<PlayerCharacter>(entityOwner.Owner, out var playerCharacter) ||
                    !entityManager.TryGetComponentData<User>(playerCharacter.UserEntity, out var user))
                {
                    continue;
                }
                
                Helper.ApplyBuff(playerCharacter.UserEntity, entityOwner, Helper.AppliedBuff);
            }
        }
    }
    
    [HarmonyPostfix]
    [HarmonyPatch(typeof(WeaponLevelSystem_Destroy), nameof(WeaponLevelSystem_Destroy.OnUpdate))]
    private static void WeaponDestoryPostfix(WeaponLevelSystem_Destroy __instance)
    {
        if (Plugin.ShouldApplyBuffs)
        {
            Plugin.Log(LogSystem.Buff, LogLevel.Info, "WeaponLevelSystem spawn POST");
            var entityManager = __instance.EntityManager;
            var entities = __instance.__query_1111682408_0.ToEntityArray(Allocator.Temp);
            foreach (var entity in entities)
            {
                if (!entityManager.TryGetComponentData<EntityOwner>(entity, out var entityOwner) ||
                    !entityManager.TryGetComponentData<PlayerCharacter>(entityOwner.Owner, out var playerCharacter) ||
                    !entityManager.TryGetComponentData<User>(playerCharacter.UserEntity, out var user))
                {
                    continue;
                }
                
                Helper.ApplyBuff(playerCharacter.UserEntity, entityOwner, Helper.AppliedBuff);
            }
        }
    }
```

In `Helper.GetAllStatBonuses`, uncomment the `WeaponMasterySystem` line.
This will ensure that anytime a mastery tick occurs, the buff is updated with the new value.

Change the `WeaponMasterySystem.BuffReceiver` to use this code instead:
Note that `weaponMastery.Mastery` is already a percentage (ie, between 0 and 1) so you don't need to divide by 100. You do also need to multiply by effectiveness to get the multiple levels of 100% mastery.

```

            var masteryType = WeaponToMasteryType(GetWeaponType(owner, out var weaponEntity));
            var weaponMasterData = Database.PlayerWeaponmastery[steamID];
            
            if (_em.TryGetBuffer<ModifyUnitStatBuff_DOTS>(weaponEntity, out var statBuffer))
            {
                var weaponMastery = weaponMasterData[masteryType];
                foreach (var statModifier in statBuffer)
                {
                    if (statModifier.StatType.IsOffensiveStat())
                    {
                        var statsIncrease = weaponMastery.Mastery * weaponMastery.Effectiveness * (OffensiveStatIncreaseFactor / 100);
                        statBonus[statModifier.StatType] += statModifier.Value * (float)statsIncrease;
                    }
                }
            }
```

This change might be a good stop-gap measure until there is a better version tying all the systems together. This doesn't take into account spell mastery or any mastery other than the currently equipped weapon. It seriously discourages people from changing weapons (as any change to a different weapon will lose their massive mastery bonus on their main weapon).

Do you have any ideas about how you would tie this to a wider system designed to support user choice and to encourage experimentation with weapons/spells? How does this support the spell mastery?

nerzhei committed:
addressed comments

nerzhei committed:
addressed comments

nerzhei commented:
We can introduce a primary weapon system where users specify the primary weapon, and its buffs apply to the current weapon at a reduced rate. Alternatively, a simpler solution would be to apply all mastery buffs, excluding those from the current weapon, at a reduced rate as well.

This does not support spell mastery, I'm working on it.

nerzhei committed:
added spell mastery buffs, added command to print detailed mastery info

nerzhei committed:
removed unused usings

aontas commented on Configuration/MasteryConfig.cs:
```diff
@@ -33,5 +33,12 @@ public static void Initialize()
         WeaponMasterySystem.EffectivenessSubSystemEnabled = _configFile.Bind("Mastery", "Enable Effectiveness Subsystem", true, "Enables the Effectiveness mastery subsystem, which lets you reset your mastery to gain a multiplier to the effects of the matching mastery.").Value;
         WeaponMasterySystem.GrowthPerEffectiveness = _configFile.Bind("Mastery", "Growth per effectiveness", 1.0, "Used to determine how mastery growth is affected at higher effectiveness levels. When enabled, more effectiveness will mean that it is slower to gain mastery (at 1, 200% effectiveness gives a growth rate of 50%). A smaller number here will reduce the growth rate faster. Use a value o ...
         WeaponMasterySystem.MaxEffectiveness = _configFile.Bind("Mastery", "Maximum Effectiveness", 5.0, "The maximum mastery effectiveness where 1 is 100%.").Value;
+        WeaponMasterySystem.OffensiveStatIncreaseFactor = _configFile.Bind("Mastery", "Offensive Stat Increase Factor", 10, "The percentage increase in offensive stats for every 100% mastery").Value;
+
+        WeaponMasterySystem.AdditionalSpellPowerPer100PercentMastery = _configFile.Bind("Spell Mastery", "Additional Spell Power per 100% Mastery", 10, "The additional amount of spell power gained per 100% mastery.").Value;
+        WeaponMasterySystem.AdditionalSpellCriticalStrikeChancePer100PercentMastery = _configFile.Bind("Spell Mastery", "Additional Spell Critical Strike Chance per 100% Mastery", 5, "The additional amount of spell critical strike chance gained per 100% mastery.").Value;
+        WeaponMasterySystem.AdditionalSpellCriticalStrikeDamagePer100PercentMastery = _configFile.Bind("Spell Mastery", "Additional Spell Critical Strike Damage per 100% Mastery", 25, "The additional amount of spell critical strike damage gained per 100% mastery.").Value;
+        WeaponMasterySystem.AdditionalSpellLifeLeechPer100PercentMastery = _configFile.Bind("Spell Mastery", "Additional Spell Life Leech per 100% Mastery", 3, "The additional amount of spell life leech gained per 100% mastery.").Value;
+        WeaponMasterySystem.AdditionalSpellCooldownRecoveryRatePer100PercentMastery = _configFile.Bind("Spell Mastery", "Additional Spell Cooldown Recovery Rate per 100% Mastery", 10, "The additional amount of spell cooldown recovery rate gained per 100% mastery.").Value;
```
If you want to go down the route of this configuration, then I would suggest it would be better to use the existing `Database.MasteryStatConfig` so we don't have to clutter the config with large lists of basically the same config.

You can set the default rates of improvement for each weapon type in the `WeaponMasterySystem.DefaultMasteryConfig`, which then gets written out to `Data\weaponMasteryConfig.json` and is user editable.

This default config is essentially using the following config:
```
{ MasteryType.Spell, new List<StatConfig>()
                {
                    new(UnitStatType.SpellPower, 0,  0.1f ),
                    new(UnitStatType.SpellCriticalStrikeChance, 0,  0.05f ),
                    new(UnitStatType.SpellCriticalStrikeDamage, 0,  0.25f ),
                    new(UnitStatType.SpellLifeLeech, 0,  0.03f ),
                    new(UnitStatType.SpellCooldownRecoveryRate, 0,  0.1f )
                } }
```

nerzhei replied:
I think this is a good approach because, as far as I can see, we don't have a weapon or something similar to buff regarding spell power. I've modified my solution to use the masteryStatConfig.

aontas commented on Systems/WeaponMasterySystem.cs:
```diff
@@ -97,12 +105,23 @@ public static void UpdateMastery(ulong steamID, MasteryType masteryType, double
+                else
+                {
+                    var message =
+                        L10N.Get(L10N.TemplateKey.MasteryFull)
```
Nice!

aontas commented:
The way I updated the `BuffReceiver` in the WeaponMasterySystem (as well as the corresponding code in BloodlineMasterySystem) dies work in that way of providing the currently equipped weapon with max buff based on the config stats, and every other weapon mastery supplied a buff at a reduced value. If you reverted your changes to `BuffReceiver` and just updated the config file then it would likely do most of what you are talking about here.

It doesn't buff the stats based on the stats of the equipped weapon, but rather based on the configured stats (so you don't have to use the best weapon in the game to make sure you are getting the best buffs). Using the equipped weapon buffs might get a bit wonky late game with they have all sorts of other stat buffs (not just physical power).

aontas commented on Commands/MasteryCommands.cs:
```diff
@@ -46,6 +46,20 @@ public static void GetMastery(ChatCommandContext ctx, string weaponType = "")
             ctx.Reply(GetMasteryDataStringForType(type, data));
         }
+        
+        [Command("details", "d", "", "Displays detailed information about benefits from mastery for your equipped weapon")]
```
This can be OK for providing more specific information solely on the weapon buffs, but you can view this same information (along with the HP buff from XP) by using `.playerinfo`. I'd probably lean towards providing a consolidated stat display command there so that as this (and the other buff systems) come online that the user can see all their buffs in one place rather than needing to type 1 command for each system.

aontas commented on Commands/MasteryCommands.cs:
```diff
@@ -90,6 +104,24 @@ private static string GetMasteryDataStringForType(WeaponMasterySystem.MasteryTyp
+
+            foreach (var line in weaponMasteryPrintableData)
+            {
+                ctx.Reply(line.Build(language));
```
If really want to have a dedicated print for mastery, you should use `Output.ChatReply(ctx, weaponMasteryPrintableData.ToArray());` as that should format it better without adding in all the extra `[SYSTEM]` parts of the reply.

nerzhei replied:
Maybe we can keep it until we improve the player info command. However, printing too much info makes it very difficult to read. I moslty made this command for debug because is difficult to see the buffs within the game.

nerzhei committed:
fixed mastery print

nerzhei committed:
updated spell mastery settings

nerzhei committed:
added resource to buff condition

nerzhei commented:
I've added `statModifier.StatType.IsResourceStat()` to the buff condition because resource farming also depends on weapon type and thus mastery type, so I think it should be buffed.

nerzhei commented:
> The way I updated the `BuffReceiver` in the WeaponMasterySystem (as well as the corresponding code in BloodlineMasterySystem) dies work in that way of providing the currently equipped weapon with max buff based on the config stats, and every other weapon mastery supplied a buff at a reduced value. If you reverted your changes to `BuffReceiver` and just updated the config file then it would likely do most of what you are talking about here.
> 
> It doesn't buff the stats based on the stats of the equipped weapon, but rather based on the configured stats (so you don't have to use the best weapon in the game to make sure you are getting the best buffs). Using the equipped weapon buffs might get a bit wonky late game with they have all sorts of other stat buffs (not just physical power).

We can go this way. However, I don't like the idea of having a flat buff. Endgame weapons have different combinations of stats, so it would be awesome to buff that specific weapon based on your mastery. Otherwise, there's no difference between having a bone sword or a dark silver sword, the benefit will be the same. A buff based on your weapon would also push players to get better sets with good stats to be enhanced by their mastery level.

aontas commented:
The benefit from mastery might be the same (for example, as master of the sword should be able to cut you with a sword, or a straw), but stronger weapons already give you more bonuses. By improving the weapon bonuses when having more mastery, this can make the end game (where weapons get extra bonus stats RANDOMLY) into something that is much harder to balance. It also gives no incentive for any other weapon type to be used, as you already have the maximum extra bonus (and other weapons will just be super weak in comparison).

Note that I am not saying that this shouldn't be an option, but more that it might be just one of the options for mastery bonus.

One other alternative is to "remember" the last equipped weapon bonus for every weapon type and use that as the basis of applying stat bonuses. So the super awesome sword at 100% mastery will still give bonuses as you grind out your whip mastery.

aontas commented on Systems/WeaponMasterySystem.cs:
```diff
@@ -293,28 +287,24 @@ public static List<L10N.LocalisableString> GetSpellMasteryPrintableData(ulong st
+                {
+                    var increasedStatString = L10N.Get(L10N.TemplateKey.MasteryAddedStat)
+                        .AddField("{statType}", $"{Enum.GetName(config.type)}")
+                        .AddField("{value}", $"+{config.rate * 100 * spellPowerAdditionFactor:F2}%");
```
Why is this x100 when spell power is not? Why not just change the config for spell power to be 0.1 instead of 10?

nerzhei replied:
This is because spell power does not scale as a percentage. Instead, you need to specify the specific value of the increase. If I change the value in the config, I will need to perform this calculation when assigning the buff (and I prefer to omit the calculation there).

nerzhei commented:
> The benefit from mastery might be the same (for example, as master of the sword should be able to cut you with a sword, or a straw), but stronger weapons already give you more bonuses. By improving the weapon bonuses when having more mastery, this can make the end game (where weapons get extra bonus stats RANDOMLY) into something that is much harder to balance. It also gives no incentive for any other weapon type to be used, as you already have the maximum extra bonus (and other weapons will just be super weak in comparison).
> 
> Note that I am not saying that this shouldn't be an option, but more that it might be just one of the options for mastery bonus.
> 
> One other alternative is to "remember" the last equipped weapon bonus for every weapon type and use that as the basis of applying stat bonuses. So the super awesome sword at 100% mastery will still give bonuses as you grind out your whip mastery.

I also like the idea of having a primary weapon that buffs everything else. For instance, you might have a sword that deals increased damage to undead because of scythe mastery, and that would be awesome.

So, if the equipped weapon equals the "primary" weapon, apply the complete mastery buff; otherwise, apply the primary weapon buff at a reduced rate.

nerzhei committed:
added primary mastery thing

nerzhei commented:
I've added some code to illustrate my ideas on the primary mastery thing. Please feel free to comment on anything you think could be improved or must be changed.

nerzhei committed:
Merge remote-tracking branch 'origin/main' into mastery-buffs

nerzhei committed:
resolved merge conflicts

aontas commented:
This is likely no longer required as the v0.2.0 version takes most of the stuff from here and includes it with the merged mastery changes.
